packages/runtime/src/components/hooks/useMediaQuery.ts

```diff
@@ -1,33 +1,35 @@
-/* eslint-env browser */
-
-import { useState } from 'react'
-
+import { useSyncExternalStore } from 'react'
 import { DeviceOverride } from '../../prop-controllers'
 import { DEVICES, findDeviceOverride, getDeviceMediaQuery } from '../utils/devices'
-import { useIsomorphicLayoutEffect } from './useIsomorphicLayoutEffect'
 
-export function useMediaQuery<S>(responsiveValue?: Array<DeviceOverride<S>>): S | void {
-  const [deviceId, setDeviceId] = useState(DEVICES[0].id)
-  const { value } = findDeviceOverride(responsiveValue, deviceId) || {}
+const DEVICE_QUERIES: { id: string; query: string }[] = DEVICES.map(device => ({
+  id: device.id,
+  query: getDeviceMediaQuery(device).replace('@media', ''),
+}))
 
-  useIsomorphicLayoutEffect(() => {
-    if (responsiveValue == null || window == null) return () => {}
-
-    const cleanUpFns = DEVICES.map(device => {
-      const mediaQueryList = window.matchMedia(getDeviceMediaQuery(device).replace('@media', ''))
+export function useMediaQuery<S>(responsiveValue?: Array<DeviceOverride<S>>): S | void {
+  const getServerSnapshot = () => findDeviceOverride(responsiveValue, DEVICES[0].id)?.value
 
-      const listener = () => {
-        if (mediaQueryList.matches) setDeviceId(device.id)
+  function getSnapshot() {
+    const deviceId: string = DEVICE_QUERIES.reduce((matchedDevice, curr) => {
+      if (window.matchMedia(curr.query).matches) {
+        return curr.id
       }
+      return matchedDevice
+    }, DEVICE_QUERIES[0].id)
+    return findDeviceOverride(responsiveValue, deviceId)?.value
+  }
 
-      listener()
-      mediaQueryList.addListener(listener)
+  function subscribe(notify: () => void) {
```

Comment:
You want the identity of the subscribe function to be stable across renders

Reply:
Can you explain more

Reply:
https://beta.reactjs.org/reference/react/useSyncExternalStore#my-subscribe-function-gets-called-after-every-re-render